Use `ts-bridge` as build tool, set `moduleResolution` to `NodeNext`

MajorLift opened the pull request:
## Explanation

As part of our OKR for upgrading TypeScript to v5.0+ (Q2 2024 O3KR4), we are updating our core dependencies to generate builds and type declarations for both CJS and ESM.

## References

- Closes: https://github.com/MetaMask/superstruct/issues/21
- Same update made to `@metamask/utils`: https://github.com/MetaMask/utils/pull/182
- Required to update core repo's `moduleResolution` setting to `NodeNext`:
  - Blocks: https://github.com/MetaMask/core/issues/3651
 
## Changelog

### Added

- Newly exports the following types: `AnyStruct`, `EnumSchema`, `InferStructTuple`, `IsExactMatch`, `IsMatch`, `IsRecord`, `IsTuple`, `ObjectSchema`, `OmitBy`, `Optionalize`, `PickBy`, `Simplify`, `UnionToIntersection`.

MajorLift committed:
Replace `ts-up` with `ts-bridge`

MajorLift committed:
Remove `types` subdirectory from `dist`

MajorLift committed:
Remove `build:types` build script

MajorLift committed:
Set TypeScript `module`, `moduleResolution` options to `NodeNext`

MajorLift committed:
Export util types used by downstream libraries

MajorLift committed:
Set package.json type to "module"

MajorLift committed:
Add explicit file extensions to relative import paths

- required by 'nodenext' option for `moduleResolution`

mcmire commented on package.json:
```diff
@@ -43,24 +43,29 @@
   },
   "license": "MIT",
   "sideEffects": false,
+  "type": "module",
```
Do we need to make this an ESM package? I think that using `ts-bridge` and upgrading the `moduleResolution` should work without this, no?

MajorLift replied:
Reverted here: https://github.com/MetaMask/superstruct/pull/18/commits/6c4e6cdae9a38476d758036c664067e88bfe9232

mcmire commented on .eslintrc.cjs:
```diff
@@ -7,6 +7,9 @@ module.exports = {
       files: ['*.ts'],
       extends: ['@metamask/eslint-config-typescript'],
+      rules: {
+        'import/no-useless-path-segments': ['error', { noUselessIndex: false }],
```
What is this for?

MajorLift replied:
Reverted here: https://github.com/MetaMask/superstruct/pull/18/commits/220fe62cae5da33108b2be348d87d53e05365168

MajorLift committed:
Revert "Set package.json type to "module""

This reverts commit e69e43668fb900280f54aaf7e5e0f51bdd731f53.

MajorLift committed:
Revert "Add explicit file extensions to relative import paths"

This reverts commit 3bfe041f1c6a30b1ad1028a3e427bba5fae14854.

MajorLift committed:
Replace `lodash-es` with `lodash`

MajorLift committed:
Update CHANGELOG

MajorLift commented:
@mcmire I removed the `"type": "module"` field in package.json for now. There was an ESM-only dependency (`lodash-es`) that was causing an error so I just replaced it with `lodash`.

mcmire reviewed (approved):
Looks good. Checked out this branch and `yarn build` locally and can confirm that this creates a different set of files than was there before.
